File: BasicAlgorithm/src/linked_list/linked_list.py
from typing import Any
class Node:
    def __init__(self, data: Any):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.size = 0
        self.head = None
        self.tail = None
    def __is_valid_index(self, index: int) -> bool:
        return ((index >= 0) and (index < self.size))
    def __node_from_data(self, data: Any) -> Node:
        return Node(data=data)
    def get_node_with_index(self, index: int) -> Node:
        current_node: Node = self.head
        for _ in range(index):
            current_node = current_node.next
        return current_node
    def insert(self, data: Any, index: int) -> None:
        node = self.__node_from_data(data=data)
        if self.size == 0:
            self.head = node
            self.tail = node
            self.size = 1
        elif index == self.size:
            self.tail.next = node
            self.tail = node
            self.size += 1
        elif not self.__is_valid_index(index):
            raise IndexError("Index out of range")
        elif index == 0:
            node.next = self.head
            self.head = node
            self.size += 1
        else:
            pre_node = self.get_node_with_index(index-1)
            node.next = pre_node.next
            pre_node.next = node
            self.size += 1
    def delete(self, index: int)->None:
        if not self.__is_valid_index(index=index):
            raise IndexError("Index out of range")
        elif index == 0:
            node = self.head
            self.head = node.next
            self.size -= 1
            del node
        elif index == self.size - 1:
            node = self.tail
            self.tail = self.get_node_with_index(index=index-1)
            self.tail.next = None
            self.size -= 1
            del node
        else:
            pre:    Node = self.get_node_with_index(index=index)
            node:   Node = pre.next
            sub:    Node = node.next
            pre.next = sub
            self.size -= 1
            del node
    def show(self) -> None:
        current_node = self.head
        while current_node is not None:
            print(f"{current_node.data}", end='')
            if current_node.next is not None:
                print(" -> ", end='')
            current_node = current_node.next
        print("") 



def test_linked_list():
    print("=" * 50)
    print("开始测试链表实现")
    print("=" * 50)
    
    # 测试1: 创建空链表
    print("\n1. 测试空链表:")
    ll = LinkedList()
    print(f"链表大小: {ll.size}")
    ll.show()  # 应该输出空行
    
    # 测试2: 插入第一个元素
    print("\n2. 测试插入第一个元素:")
    ll.insert(10, 0)
    print(f"链表大小: {ll.size}")
    ll.show()  # 应该输出: 10
    
    # 测试3: 在末尾插入元素
    print("\n3. 测试在末尾插入:")
    ll.insert(20, 1)
    ll.insert(30, 2)
    print(f"链表大小: {ll.size}")
    ll.show()  # 应该输出: 10 -> 20 -> 30
    
    # 测试4: 在头部插入元素
    print("\n4. 测试在头部插入:")
    ll.insert(5, 0)
    print(f"链表大小: {ll.size}")
    ll.show()  # 应该输出: 5 -> 10 -> 20 -> 30
    
    # 测试5: 在中间插入元素
    print("\n5. 测试在中间插入:")
    ll.insert(15, 2)
    ll.insert(25, 4)
    print(f"链表大小: {ll.size}")
    ll.show()  # 应该输出: 5 -> 10 -> 15 -> 20 -> 25 -> 30
    
    # 测试6: 删除头部元素
    print("\n6. 测试删除头部:")
    ll.delete(0)
    print(f"链表大小: {ll.size}")
    ll.show()  # 应该输出: 10 -> 15 -> 20 -> 25 -> 30
    
    # 测试7: 删除尾部元素
    print("\n7. 测试删除尾部:")
    ll.delete(4)  # 现在size=5，索引4是最后一个
    print(f"链表大小: {ll.size}")
    ll.show()  # 应该输出: 10 -> 15 -> 20 -> 25
    
    # 测试8: 删除中间元素
    print("\n8. 测试删除中间:")
    ll.delete(1)
    print(f"链表大小: {ll.size}")
    ll.show()  # 应该输出: 10 -> 20 -> 25
    
    # 测试9: 边界情况测试 - 删除最后一个元素
    print("\n9. 测试删除最后一个元素:")
    ll2 = LinkedList()
    ll2.insert(100, 0)
    ll2.delete(0)
    print(f"链表大小: {ll2.size}")
    ll2.show()  # 应该输出空行
    print(f"head: {ll2.head}, tail: {ll2.tail}")  # 都应该为None
    
    # 测试10: 异常情况测试 - 索引越界
    print("\n10. 测试异常情况:")
    try:
        ll.insert(999, 10)  # 索引越界
        print("❌ 应该抛出异常但没有抛出")
    except IndexError as e:
        print(f"✅ 正确抛出异常: {e}")
    
    try:
        ll.delete(10)  # 索引越界
        print("❌ 应该抛出异常但没有抛出")
    except IndexError as e:
        print(f"✅ 正确抛出异常: {e}")
    
    try:
        ll.delete(-1)  # 负索引
        print("❌ 应该抛出异常但没有抛出")
    except IndexError as e:
        print(f"✅ 正确抛出异常: {e}")
    
    # 测试11: 复杂操作序列
    print("\n11. 测试复杂操作序列:")
    ll3 = LinkedList()
    # 插入多个元素
    for i in range(5):
        ll3.insert(i * 10, i)
    print("初始链表:")
    ll3.show()  # 0 -> 10 -> 20 -> 30 -> 40
    
    # 混合插入删除
    ll3.delete(2)  # 删除20
    ll3.insert(25, 2)  # 在位置2插入25
    ll3.insert(5, 0)  # 在头部插入5
    ll3.delete(4)  # 删除30
    ll3.insert(35, 4)  # 在末尾插入35
    
    print("操作后链表:")
    ll3.show()  # 应该输出: 5 -> 0 -> 10 -> 25 -> 35
    
    # 测试12: 验证链表连接正确性
    print("\n12. 验证链表连接正确性:")
    current = ll3.head
    print("遍历链表:")
    while current:
        print(f"节点数据: {current.data}, 下一个节点: {current.next.data if current.next else None}")
        current = current.next
    
    print("=" * 50)
    print("测试完成!")
    print("=" * 50)


if __name__ == '__main__':
    link = LinkedList()
    link.insert(data=1, index=0)  # 1
    link.insert(data=2, index=1)  # 1 -> 2
    link.insert(data=4, index=2)  # 1 -> 2 -> 4
    link.insert(data=3, index=2)  # 1 -> 2 -> 3 -> 4
    link.show()  # 输出: 1 -> 2 -> 3 -> 4

    test_linked_list()

File: BasicAlgorithm/src/linked_list/linked_list.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct Node{
    int data;
    struct Node *next;
} Node;

typedef struct linked_list
{
    int size;
    struct Node *head;
    struct Node *tail;
} linked_list;

// 函数声明（解决顺序问题）
void show(linked_list* list);
void free_linked_list(linked_list* list);

Node* create_node(int data) {
    Node* new_node = (Node*)malloc(sizeof(Node));
    new_node->data = data;
    new_node->next = NULL;
    return new_node;
}

void init_linked_list(linked_list* list) {
    list->size = 0;
    list->head = NULL;
    list->tail = NULL;
}

bool is_valid_index(linked_list* list, int index) {
    return ((index >= 0) && (index < list->size));
}

Node* get_node_with_index(linked_list* list, int index) {
    if (!is_valid_index(list, index)) exit(-1);
    Node* current_node = list->head;
    for (int i = 0; i < index; i++) current_node = current_node->next;
    return current_node;
}

void insert(linked_list* list, int data, int index) {
    if (index < 0 || index > list->size) {
        printf("错误: 索引 %d 超出范围 (大小: %d)\n", index, list->size);
        return;
    }
    
    Node* new_node = create_node(data);
    
    if (list->size == 0) {
        list->head = new_node;
        list->tail = new_node;
        list->size = 1;
    } else if (index == list->size) {
        list->tail->next = new_node;
        list->tail = new_node;
        list->size++;
    } else if (index == 0) {
        new_node->next = list->head;
        list->head = new_node;
        list->size++;
    } else {
        Node* prev_node = get_node_with_index(list, index - 1);
        new_node->next = prev_node->next;
        prev_node->next = new_node;
        list->size++;
    }
}

void delete(linked_list* list, int index) {
    if (!is_valid_index(list, index)) {
        printf("错误: 索引 %d 超出范围 (大小: %d)\n", index, list->size);
        return;
    }
    
    if (index == 0) {
        Node* node_to_delete = list->head;
        list->head = node_to_delete->next;
        
        if (list->size == 1) {
            list->tail = NULL;
        }
        
        free(node_to_delete);
        list->size--;
    } else if (index == list->size - 1) {
        Node* prev_node = get_node_with_index(list, index - 1);
        Node* node_to_delete = list->tail;
        
        prev_node->next = NULL;
        list->tail = prev_node;
        
        free(node_to_delete);
        list->size--;
    } else {
        Node* prev_node = get_node_with_index(list, index - 1);
        Node* node_to_delete = prev_node->next;
        Node* next_node = node_to_delete->next;
        
        prev_node->next = next_node;
        free(node_to_delete);
        list->size--;
    }
}

// 显示链表内容
void show(linked_list* list) {
    Node* current_node = list->head;
    
    while (current_node != NULL) {
        printf("%d", current_node->data);
        if (current_node->next != NULL) {
            printf(" -> ");
        }
        current_node = current_node->next;
    }
    printf("\n");
}

// 释放链表内存
void free_linked_list(linked_list* list) {
    Node* current_node = list->head;
    
    while (current_node != NULL) {
        Node* next_node = current_node->next;
        free(current_node);
        current_node = next_node;
    }
    
    list->head = NULL;
    list->tail = NULL;
    list->size = 0;
}

void test_linked_list() {
    printf("==================================================\n");
    printf("Starting Linked List Implementation Test\n");
    printf("==================================================\n");
    
    linked_list ll;
    init_linked_list(&ll);
    
    // Test 1: Create empty linked list
    printf("\n1. Testing empty linked list:\n");
    printf("List size: %d\n", ll.size);
    show(&ll);
    
    // Test 2: Insert first element
    printf("\n2. Testing first element insertion:\n");
    insert(&ll, 10, 0);
    printf("List size: %d\n", ll.size);
    show(&ll);
    
    // Test 3: Insert elements at the end
    printf("\n3. Testing insertion at the end:\n");
    insert(&ll, 20, 1);
    insert(&ll, 30, 2);
    printf("List size: %d\n", ll.size);
    show(&ll);
    
    // Test 4: Insert element at the head
    printf("\n4. Testing insertion at the head:\n");
    insert(&ll, 5, 0);
    printf("List size: %d\n", ll.size);
    show(&ll);
    
    // Test 5: Insert elements in the middle
    printf("\n5. Testing insertion in the middle:\n");
    insert(&ll, 15, 2);
    insert(&ll, 25, 4);
    printf("List size: %d\n", ll.size);
    show(&ll);
    
    // Test 6: Delete head element
    printf("\n6. Testing head deletion:\n");
    delete(&ll, 0);
    printf("List size: %d\n", ll.size);
    show(&ll);
    
    // Test 7: Delete tail element
    printf("\n7. Testing tail deletion:\n");
    delete(&ll, 4);
    printf("List size: %d\n", ll.size);
    show(&ll);
    
    // Test 8: Delete middle element
    printf("\n8. Testing middle element deletion:\n");
    delete(&ll, 1);
    printf("List size: %d\n", ll.size);
    show(&ll);
    
    // Test 9: Boundary case testing
    printf("\n9. Testing last element deletion:\n");
    linked_list ll2;
    init_linked_list(&ll2);
    insert(&ll2, 100, 0);
    delete(&ll2, 0);
    printf("List size: %d\n", ll2.size);
    show(&ll2);
    printf("head: %p, tail: %p\n", (void*)ll2.head, (void*)ll2.tail);
    
    free_linked_list(&ll2);
    
    // Test 10: Exception case testing
    printf("\n10. Testing exception cases:\n");
    insert(&ll, 999, 10);
    delete(&ll, 10);
    delete(&ll, -1);
    
    // Test 11: Complex operation sequence
    printf("\n11. Testing complex operation sequence:\n");
    linked_list ll3;
    init_linked_list(&ll3);
    
    for (int i = 0; i < 5; i++) {
        insert(&ll3, i * 10, i);
    }
    printf("Initial linked list:\n");
    show(&ll3);
    
    delete(&ll3, 2);
    insert(&ll3, 25, 2);
    insert(&ll3, 5, 0);
    delete(&ll3, 4);
    insert(&ll3, 35, 4);
    
    printf("Linked list after operations:\n");
    show(&ll3);
    
    // Test 12: Verify linked list connection correctness
    printf("\n12. Verifying linked list connection correctness:\n");
    Node* current = ll3.head;
    printf("Traversing linked list:\n");
    while (current) {
        printf("Node data: %d, Next node: ", current->data);
        if (current->next) {
            printf("%d\n", current->next->data);
        } else {
            printf("NULL\n");
        }
        current = current->next;
    }
    
    free_linked_list(&ll3);
    free_linked_list(&ll);
    
    printf("==================================================\n");
    printf("Testing completed!\n");
    printf("==================================================\n");
}

int main(void) {
    // 基本测试
    linked_list link;
    init_linked_list(&link);
    
    insert(&link, 1, 0);
    insert(&link, 2, 1);
    insert(&link, 4, 2);
    insert(&link, 3, 2);
    
    printf("Test begins:\n");
    show(&link);
    
    free_linked_list(&link);
    
    // 运行完整测试
    test_linked_list();
    
    return 0;
}